### PR TITLE
add letsencrypt to single docker image

### PR DESCRIPTION
**Pull request by jonnymccullagh**
## Description
- added letsencrypt folder with nginx snippets and shell script to be triggered if the CUSTOM_DOMAIN variable is specified
- Amended nginx conf to use standard locations (/var/log, /var/run) and username (www-data)
- Added healthcheck for docker container
- Added README with instructions for building the single image

### CODE REVIEW

File: hosting/scripts/healthcheck.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+healthy=true
+
+if [[ $(curl -Lfk -s -w "%{http_code}\n" http://localhost/ -o /dev/null) -ne 200 ]]; then
```
Comment (mike12345567) [inline hosting/scripts/healthcheck.sh:4]: Love this great idea for confirming everything working as expected!

File: hosting/letsencrypt/ssl-dhparams.pem [hunk 2]
```diff
@@ -0,0 +1,8 @@
+-----BEGIN DH PARAMETERS-----
```
[the remark this answers is not in the document]
Comment (jonnymccullagh) [reply hosting/letsencrypt/ssl-dhparams.pem:1]: Generating a new one takes a very long time and based on a search 
"We don't have any reason to think that 2048-bit DL is tractable, even if
the parameters are widely shared."
Also certbot distribute the dhparams file in their own repo.

**Review by mike12345567 — APPROVED** [review]
LGTM!

**Comment by aptkingston** [comment]
Nice one @jonnymccullagh! 2 suggestions after looking over the files:

1. When lets encrypt is enabled and we're using `nginx-ssl.conf`, we're still using the same base `nginx.conf` which is listening on port 80, so we won't be automatically redirecting to 443. Should be consider adding a 301 redirect on port 80 to redirect to 443 instead to ensure we're always using TLS when it's enabled?
2. The full proxy conf is defined in both those files, so could we consider pulling out the core proxy config into it's own conf file and including that in both the TLS and non-TLS configs? Might make things easier for maintenance down the line to only have one config to keep up to date

**Comment by jonnymccullagh** [comment]
@aptkingston Yes I agree that it could be more efficient but I didn't want to assume the use of https. The user may not wish to use https/certbot/port 443. They may terminate SSL at another proxy in front (e.g. traefik,nginx proxy manager) to proxy to port 80. So they could avoid using certbot altogether. So I was aiming for making port 80 work regardless and add port 443 if things work out ok. Otherwise I feel like we would need a parameter for whether to FORCE_SSL but if anything happens to the cert process we would be redirecting to a cert warning page.
I don't think there is a right answer.

**Review by shogunpurple — APPROVED** [review]
This is excellent - great work on this. Especially some of the other additional stuff like the healthchecks and cert renewals 👌

**Comment by aptkingston** [comment]
> @aptkingston Yes I agree that it could be more efficient but I didn't want to assume the use of https. The user may not wish to use https/certbot/port 443. They may terminate SSL at another proxy in front (e.g. traefik,nginx proxy manager) to proxy to port 80. So they could avoid using certbot altogether. So I was aiming for making port 80 work regardless and add port 443 if things work out ok. Otherwise I feel like we would need a parameter for whether to FORCE_SSL but if anything happens to the cert process we would be redirecting to a cert warning page. I don't think there is a right answer.

Yea I just mean if they have the `CUSTOM_DOMAIN` env variable set then they are planning for the termination done at the Budibase level - but I get your point that they might prefer the option to use port 80 in the event that cert renewal expires. Likewise there maybe be situations that they would rather see a cert error than allow insecure connections, so I agree that there's no right answer without letting the user decide.

Wouldn't worry about making any changes for now then.

**Comment by jonnymccullagh** [comment]
ok let's see how people get on when they deploy

**Pull request merged**